Extract `ImpSpec` into its own package

lehins opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins committed:
Extract reusable Imp test functionality into `ImpSpec` package

This required some slight changes to the interface, which turned out to
be an overall improvemenoverall.

Soupstraw commented on libs/cardano-ledger-api/test/Tests.hs:
```diff
@@ -25,9 +25,10 @@ apiSpec =
-        withImpStateWithProtVer @Conway v $ do
-          ImpQuery.spec @Conway
+      withImpInit @(LedgerSpec Conway) $
+        forM_ [eraProtVerLow @Conway .. eraProtVerHigh @Conway] $ \v ->
```
I've seen this pattern in a couple of places, maybe this should be a function

lehins replied:
Added a reusable `eraProtVersions` function.

Lucsanszky commented on libs/ImpSpec/stack.yaml:
Do we want to include this? We don't include `stack.yaml` files anywhere else.

lehins replied:
No, this is not gonna be included. I only used it locally for my own development. In fact I'll add it to `.gitignore` just in case.

Lucsanszky reviewed (approved):
This is great! Apart from a minor thing, all looks good. :+1:

Soupstraw commented on libs/ImpSpec/src/Test/ImpSpec/Internal.hs:
```diff
@@ -0,0 +1,388 @@
+      , ["Failed with Exception:", indent 4 $ pretty (displayException e)]
+      ]
+
+newtype ImpM t a = ImpM {unImpM :: ReaderT (ImpEnv t) IO a}
```
An alternative way to do this would've been to parametrize `ImpM` by `ImpSpecEnv` and  `ImpSpecState` directly instead of using this index `t`.

lehins replied:
It seems at first that it work, but:
* it would lead to terrible type signatures on all functions (i.e. `impTest :: ImpM (ImpLedgerEnv era) (ImpLedgerState era) ()`).
* it would not be as future extensible as current approach (eg. adding potentially other things akin to environment or state, like events of some sorts for instance)
* it would make the whole initialization much uglier:
```haskell
class ImpSpec e s where

  impInitIO :: QCGen -> IO (ImpInit e s t)
```
* Most importantly we would not have a way to implement this kind of functionality:
```haskell
class ImpSpec t where
  impPrepAction :: ImpM t ()
```
  In other words there would be no way to customize Imp functionality that does not depend on neither state nor environment.

Those are just facts that immediately come to my mind, but in general for any Haskell interface, if it is possible to reduce number of type parameters then it is a good idea to do so.

lehins committed:
Add `stack.yaml` to `.gitignore`

We no longer use `stack`, but it can sometimes be useful for development
locally

lehins committed:
Add `eraProtVersions`

neilmayhew commented on eras/mary/impl/CHANGELOG.md:
```diff
@@ -1,9 +1,13 @@
+
+* Switch to using `ImpSpec` package
+
 ## 1.7.0.1
 
 *
```
Should these lines have been taken out?

lehins replied:
`1.7.0.1` have been released to CHaP, so removing the whole section would be wrong.

The fact that there is an empty bullet point in that section is not a problem, quite the opposite, it serves as a good indicator that there were no notable changes in that release. In fact if we follow our release process we can't remove such empty bullet points, because that implies a PR before making a release, which we do not require.
We could remove it after the release, but that would make little difference, since released version of the package would still have this empty bullet point in its changelog.

So, I don't think we need to worry about this.

neilmayhew replied:
Thanks for the explanation. That makes sense. I had been wondering why we made a release when there was nothing to note in the changelog, but I was forgetting that the changelog is for externally-visible changes only.
